### PR TITLE
Memory overcommit: continue query execution if memory is available

### PR DESCRIPTION
**Pull request by novikd**
Changelog category (leave one):

- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

If the required amount of memory is available before the selected query stopped, all waiting queries continue execution. Now we don't stop any query if memory is freed before the moment when the selected query knows about the cancellation.

### CODE REVIEW

**Comment by novikd** [comment]
@Mergifyio update

**Comment by alesapin** [comment]
@mergifyio update

**Comment by alesapin** [comment]
Looks like it helped, let's try one more time.

**Comment by alesapin** [comment]
@Mergifyio update

File: src/Common/OvercommitTracker.cpp [hunk 1]
```diff
@@ -36,42 +39,82 @@ bool OvercommitTracker::needToStopQuery(MemoryTracker * tracker)
...
     });
-    if (timeout)
-        LOG_DEBUG(getLogger(), "Need to stop query because reached waiting timeout");
-    else
-        LOG_DEBUG(getLogger(), "Memory freed within timeout");
-    return timeout;
+    LOG_DEBUG(getLogger(), "Memory was{} freed within timeout", (timeout ? " not" : ""));
+
+    required_memory -= amount;
+    auto still_need = required_per_thread[tracker]; // If enough memory is freed it will be 0
```
Comment (alesapin) [inline src/Common/OvercommitTracker.cpp:71]: bad `auto`

File: src/Common/OvercommitTracker.cpp [hunk 2]
```diff
@@ -36,42 +39,82 @@ bool OvercommitTracker::needToStopQuery(MemoryTracker * tracker)
...
+    required_per_thread.erase(tracker);
+
+    // If threads where not released since last call of this method,
+    // we can release them now.
+    if (allow_release && required_memory <= freed_momory && still_need != 0)
```
Comment (alesapin) [inline src/Common/OvercommitTracker.cpp:76]:
```suggestion
    if (allow_release && required_memory <= freed_memory && still_need != 0)
```

File: src/Common/OvercommitTracker.cpp [hunk 3]
```diff
@@ -11,8 +11,11 @@ constexpr std::chrono::microseconds ZERO_MICROSEC = 0us;
 OvercommitTracker::OvercommitTracker(std::mutex & global_mutex_)
     : max_wait_time(ZERO_MICROSEC)
     , picked_tracker(nullptr)
-    , cancelation_state(QueryCancelationState::NONE)
+    , cancellation_state(QueryCancellationState::NONE)
     , global_mutex(global_mutex_)
+    , freed_momory(0)
```
Comment (alesapin) [inline src/Common/OvercommitTracker.cpp:16]:
```suggestion
    , freed_memory(0)
```

File: src/Common/OvercommitTracker.h [hunk 4]
```diff
@@ -58,43 +67,54 @@ struct OvercommitTracker : boost::noncopyable
...
+        cancellation_state = QueryCancellationState::NONE;
+        freed_momory = 0;
+        allow_release = true;
+    }
+
+    void releaseThreads();
+
+    QueryCancellationState cancellation_state;
+
+    std::unordered_map<MemoryTracker *, Int64> required_per_thread;
```
Comment (alesapin) [inline src/Common/OvercommitTracker.h:106]: Why Int64? We can require negative amount of memory?

**Review by alesapin — APPROVED** [review]

File: src/Common/OvercommitTracker.h (hunk 4 above)
Comment (novikd) [reply src/Common/OvercommitTracker.h:106]: Because interface of `MemoryTracker` receives `Int64` in both methods `alloc` and `free`.

**Comment by alesapin** [comment]
@Mergifyio update

**Comment by novikd** [comment]
> Style Check (actions) — Check style failed.

Unrelated.

> Stress test (debug, actions) — Backward compatibility check: Failed to download previous release

Unrelated. Fatal messages fixed in #36660

> Stress test (address, actions) — Hung check failed
> Stress test (memory, actions) — Hung check failed

`DROP DATABASE test` and `CREATE DATABASE IF NOT EXISTS test` hung.

> Stress test (thread, actions) — Backward compatibility check: Error message
> Stress test (undefined, actions) — Backward compatibility check: Error message

Unclear problems, need to take a look

**Comment by novikd** [comment]
> Stress test (thread, actions) — Backward compatibility check: Error message
> Stress test (undefined, actions) — Backward compatibility check: Error message

Should be fixed in #36845

**Pull request merged**